### PR TITLE
EIP 1283 SSTORE

### LINKED ISSUE #354: [Constantinople] EIP 1283 SSTORE net metering
[EIP 1283](https://eips.ethereum.org/EIPS/eip-1283) SSTORE net metering without dirty maps (replaces EIP-1087)

Test Situation:
Open PR, affecting lots of tests
https://github.com/ethereum/tests/pull/511
https://github.com/ethereum/tests/issues/483

Implementation in the VM:
Mostly/all in the [SSTORE](https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/opFns.js#L362) opcode code.

### PR DESCRIPTION
**Pull request by vpulim**
Depends on ethereumjs/ethereumjs-common#27.

Closes #354

I wasn't sure about the best way to test this PR since there aren't any EIP-1283 tests yet in ethereum/tests, so I created my own tests in ``tests/constantinopleSstoreTest.js`` until official tests are available. These tests are currently not being run in circleci. Ideas?

### CODE REVIEW

**Comment by holgerd77** [comment]
If I got this correct I think the latest PR merge on the tests repository https://github.com/ethereum/tests/pull/511 should contain various tests (primarily testing other functionality) implicitly testing the new gas prices. This is just three days old, we'll have to do a new release with an updated submodule of our own proxy testing module (``ethereumjs-testing``).

**Comment by vpulim** [comment]
@holgerd77 Ah ok, great. I didn't realize they were just released.

**Comment by holgerd77** [comment]
Ok, I just released a new version of the ``ethereumjs-common`` library [v0.6.0](https://github.com/ethereumjs/ethereumjs-common/releases/tag/v0.6.0), so this can be updated now.

**Comment by holgerd77** [comment]
Update: I would update the ``ethereumjs-common`` dependency myself and then do a review and eventually merge if you don't intervene in the next hour or so, then we can build/rebase other stuff on top.

Side note: I thought of the labels ``PR state: WIP`` and ``PR state: needs review`` somewhat excluding each other, so the WIP semantics would be more of "someone is still actively working on it" explicitly leaving out the review as part of the work. With this notion I think it is more useful, then one can switch - once work is ready - from WIP to review label and it's more clear for others what to do. Since you assigned both labels simultaneously I assume that you have done this with the code-work-side of this PR ready and therewith open for a review.

**Comment by vpulim** [comment]
It looks like #369 is causing these errors. If that is expected behavior due to a dependency on other open PRs, then please proceed with the review/merge.

File: lib/opFns.js [hunk 1]
```diff
@@ -1024,3 +1015,73 @@ function makeCall (runState, callOptions, localOpts, cb) {
...
+      // If original value equals new value (this storage slot is reset)
+      if (original.length === 0) {
+        // If original value is 0, add 19800 gas to refund counter.
+        runState.gasRefund = runState.gasRefund.addn(runState._common.param('gasPrices', 'netSstoreResetClearRefund'))
+      } else {
+        // Otherwise, add 4800 gas to refund counter.
+        runState.gasRefund = runState.gasRefund.addn(runState._common.param('gasPrices', 'netSstoreResetRefund'))
+      }
+    }
+    return subGas(runState, new BN(runState._common.param('gasPrices', 'netSstoreDirtyGas')))
```
Comment (holgerd77) [inline lib/opFns.js]: This is working down the various cases from the EIP and adding to the existing ``runState.gasRefund`` variable. I assume this is correct since tests are passing here on a broad basis.

If there are smaller caveats or edge cases we can do subsequent PRs to correct once we have a clearer picture on the test runs and tests have stabilized more.

**Comment by holgerd77** [comment]
Hi Vinay, have now merged, thanks a lot that you took on this so quickly! 😄 

Hmm, the new test failure caused by #369 wasn't intended, but we can then probably address this separately in a new PR, if you want you can file a new issue for that. No need to wait here since #369 is already merged anyhow and we will not do an imminent release on the VM.

**Comment by holgerd77** [comment]
Short test result summary (ethereumjs-testing v1.2.3, submodule 9777827):

Before:

```
1..4776
# tests 4776
# pass  3783
# fail  993
```

After:

```
1..4776
# tests 4776
# pass  4684
# fail  92
```

Really cool! 🏆 🏆
And some tropical drink for the win: 🍹 😄

File: lib/opFns.js [hunk 2]
```diff
@@ -1024,3 +1015,73 @@ function makeCall (runState, callOptions, localOpts, cb) {
...
+
+function updateSstoreGas (runState, found, value) {
+  if (runState._common.gteHardfork('constantinople')) {
+    var original = found.original
+    var current = found.current
+    if (current.equals(value)) {
+      // If current value equals new value (this is a no-op), 200 gas is deducted.
+      subGas(runState, new BN(runState._common.param('gasPrices', 'netSstoreNoopGas')))
+      return
+    }
```
[the remark this answers is not in the document]
Comment (vpulim) [reply lib/opFns.js]: Good catch! This should be ``return subGas...`` like in the rest of the code. This is stylistic and can be fixed later... not critical.

**Comment by vpulim** [comment]
Woo hoo! 🙌
